bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -2,454 +2,172 @@
-			// Change origin to the bridge.
-			UniversalOrigin(GlobalConsensus(network)),
-			// Call create_asset on foreign assets pallet.
-			Transact {
-				origin_kind: OriginKind::Xcm,
-				call: (
-					create_call_index,
-					asset_id,
-					MultiAddress::<[u8; 32], ()>::Id(owner),
-					MINIMUM_DEPOSIT,
-				)
-					.encode()
-					.into(),
-			},
-			// Forward message id to Asset Hub
-			SetTopic(message_id.into()),
-			// Once the program ends here, appendix program will run, which will deposit any
-			// leftover fee to snowbridge sovereign.
-		]
-		.into();
+	fn convert(message: Message) -> Result<Xcm<()>, ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			// Decode xcm
+			let versioned_xcm = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			)
+			.map_err(|_| ConvertMessageError::InvalidVersionedXCM)?;
+			message_xcm = versioned_xcm.try_into().map_err(|_| ConvertMessageError::InvalidXCM)?;
+		}
 
-		(xcm, total_amount.into())
-	}
+		log::debug!(target: LOG_TARGET,"xcm decoded as {:?}", message_xcm);
 
-	fn convert_send_token(
-		message_id: H256,
-		chain_id: u64,
-		token: H160,
-		destination: Destination,
-		amount: u128,
-		asset_hub_fee: u128,
-	) -> (Xcm<()>, Balance) {
-		let network = Ethereum { chain_id };
-		let asset_hub_fee_asset: Asset = (Location::parent(), asset_hub_fee).into();
-		let asset: Asset = (Self::convert_token_address(network, token), amount).into();
+		let network = EthereumNetwork::get();
 
-		let (dest_para_id, beneficiary, dest_para_fee) = match destination {
-			// Final destination is a 32-byte account on AssetHub
-			Destination::AccountId32 { id } =>
-				(None, Location::new(0, [AccountId32 { network: None, id }]), 0),
-			// Final destination is a 32-byte account on a sibling of AssetHub
-			Destination::ForeignAccountId32 { para_id, id, fee } => (
-				Some(para_id),
-				Location::new(0, [AccountId32 { network: None, id }]),
-				// Total fee needs to cover execution on AssetHub and Sibling
-				fee,
-			),
-			// Final destination is a 20-byte account on a sibling of AssetHub
-			Destination::ForeignAccountId20 { para_id, id, fee } => (
-				Some(para_id),
-				Location::new(0, [AccountKey20 { network: None, key: id }]),
-				// Total fee needs to cover execution on AssetHub and Sibling
-				fee,
-			),
-		};
+		let origin_location = Location::new(2, GlobalConsensus(network))
+			.push_interior(AccountKey20 { key: message.origin.into(), network: None })
+			.map_err(|_| ConvertMessageError::InvalidXCM)?;
 
-		let total_fees = asset_hub_fee.saturating_add(dest_para_fee);
-		let total_fee_asset: Asset = (Location::parent(), total_fees).into();
-		let inbound_queue_pallet_index = InboundQueuePalletInstance::get();
+		let network = EthereumNetwork::get();
 
+		let fee_asset = Location::new(1, Here);
+		let fee_value = 1_000_000_000u128; // TODO get from command
```

Comment:
> Should make the fee configurable via runtime config, ie add T::XcmPrologueFee: Balance pallet config

Can we make this fee configurable? Iirc we cannot, because it could include arbitrary instructions such as Transact. I explained the flow as I see it should work over here: https://github.com/claravanstaden/polkadot-sdk/pull/3#discussion_r1846293603

Reply:
That's an interesting idea @yrong. It could work. I think the main downside to this (from my perspective) is that many governance calls will be a pain to do. It will require more invention from our team's side, more coordination with other teams.

What is your concern with the dry-run approach? Isn't runtime API calls free so there's no cost to doing the estimation?

Reply:
> I still think InboundMessage should contain the fee like V1 before but with AssetId including.

Yeah, that's why I still prefer this.

Reply:
I think there's a misunderstanding here how fees are supposed to be paid on AH. Please read the [API docs](https://github.com/Snowfork/snowbridge/blob/553b0b9d9ec6f3683c6c605417ea5532282771ed/contracts/src/v2/IGateway.sol#L54) for `IGateway.sendMessage`:

>     // The `xcm` should contain the necessary instructions to:
>     // 1. Pay XCM execution fees for `xcm`, either from assets in holding,
>     //    or from the sovereign account of `msg.sender`.
>     // 2. Handle the assets in holding, either depositing them into
>     //    some account, or forwarding them to another destination.

Or in other words, its the users responsibility to provide assets and xcm instructions to cover fees.

Its only the fixed xcm _prologue_ that requires a small fixed fee to be teleported by the relayer account on BH.

Our governance messages for registering ENA tokens will use the same API.

Reply:
I see, thanks for clearing it up @vgeddes. I will add it.